ui/appui-layout-react/src/appui-layout-react/state/WidgetState.ts

```diff
@@ -0,0 +1,105 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Base
+ */
+
+import { UiError } from "@itwin/appui-abstract";
+import { RectangleProps } from "@itwin/core-react";
+import produce from "immer";
+import { PanelSide } from "../widget-panels/Panel";
+import { category } from "./internal/NineZoneStateHelpers";
+import { addWidgetState, createFloatingWidgetState, createPopoutWidgetState } from "./internal/WidgetStateHelpers";
+import { NineZoneState } from "./NineZoneState";
+import { TabState } from "./TabState";
+
+/** State of a stacked widget, which can contain multiple tabs. I.e. in a panel section or a floating widget.
+ * @internal
+ */
+export interface WidgetState {
+  readonly activeTabId: TabState["id"];
+  readonly id: string;
+  readonly minimized: boolean;
+  readonly tabs: ReadonlyArray<TabState["id"]>;
+  readonly isFloatingStateWindowResizable?: boolean;
+}
+
+/** @internal */
+export interface WidgetsState { readonly [id: string]: WidgetState }
+
+/** @internal */
+export interface FloatingWidgetHomeState {
+  readonly widgetIndex: number;
+  readonly widgetId: WidgetState["id"] | undefined;
+  readonly side: PanelSide;
+}
+
+/** @internal */
+export interface FloatingWidgetState {
+  readonly bounds: RectangleProps;
+  readonly id: WidgetState["id"];
+  readonly home: FloatingWidgetHomeState;
+  readonly userSized?: boolean;
+  readonly hidden?: boolean;
+}
+
+/** @internal */
+export interface FloatingWidgetsState {
+  readonly byId: { readonly [id: string]: FloatingWidgetState };
+  readonly allIds: ReadonlyArray<FloatingWidgetState["id"]>;
+}
+
+/** @internal */
+export interface PopoutWidgetState {
+  readonly bounds: RectangleProps;
+  readonly id: WidgetState["id"];
+  // TODO: popout widget home could also be a floating widget.
+  readonly home: FloatingWidgetHomeState;
+}
+
+/** @internal */
+export interface PopoutWidgetsState {
+  readonly byId: { readonly [id: string]: PopoutWidgetState };
+  readonly allIds: ReadonlyArray<PopoutWidgetState["id"]>;
+}
+
+/** @internal */
+export function addFloatingWidget(state: NineZoneState, id: FloatingWidgetState["id"], tabs: WidgetState["tabs"], floatingWidgetArgs?: Partial<FloatingWidgetState>,
```

Comment:
I've avoided that to minimize the things that are exported in a barrel file, since `removeFloatingWidget` is never called in **appui-react**. `addFloatingWidget` is called to initialize a floating widget, but after that it is managed by adding/removing the tabs.